src/state/zustand/swapperStore/actions.ts

```diff
@@ -1,62 +1,175 @@
+        draft.buyAmountFiat = '0'
+        draft.sellAmountFiat = '0'
+        return draft
       },
+      false,
+      `swapper/clearAmounts`,
     )
 
-export const toggleIsExactAllowance =
-  (set: SetSwapperStoreAction<SwapperStore>): SwapperAction['toggleIsExactAllowance'] =>
+export const handleSwitchAssets =
+  (set: SetSwapperStoreAction<SwapperState>): SwapperState['handleSwitchAssets'] =>
   () =>
     set(
-      state => {
-        state.isExactAllowance = !state.isExactAllowance
-        return state
+      draft => {
+        const sellAsset = draft.sellAsset
+        const buyAsset = draft.buyAsset
+        const sellAssetFiatRate = draft.sellAssetFiatRate
+        const buyAssetFiatRate = draft.buyAssetFiatRate
+
+        draft.buyAsset = sellAsset
+        draft.sellAsset = buyAsset
+        draft.sellAmountCryptoPrecision = '0'
+        draft.buyAmountCryptoPrecision = '0'
+        draft.sellAmountFiat = '0'
+        draft.buyAmountFiat = '0'
+        draft.amount = '0'
+        draft.feeAssetFiatRate = undefined
+        draft.fees = undefined
+        draft.trade = undefined
+        draft.selectedSellAssetAccountId = undefined
+        draft.selectedBuyAssetAccountId = undefined
+        draft.buyAssetAccountId = undefined
+        draft.sellAssetAccountId = undefined
+        draft.buyAssetFiatRate = sellAssetFiatRate
+        draft.sellAssetFiatRate = buyAssetFiatRate
+        return draft
       },
       false,
-      `swapper/toggleIsExactAllowance`,
+      `swapper/handleSwitchAssets`,
     )
 
-export const clearAmounts =
-  (set: SetSwapperStoreAction<SwapperStore>): SwapperAction['clearAmounts'] =>
+export const handleInputAmountChange =
+  (set: SetSwapperStoreAction<SwapperState>): SwapperState['handleInputAmountChange'] =>
   () =>
     set(
-      state => {
-        state.sellAmountCryptoPrecision = '0'
-        state.buyAmountCryptoPrecision = '0'
-        state.buyAmountFiat = '0'
-        state.sellAmountFiat = '0'
-        state.amount = '0'
-        state.isSendMax = false
-        state.action = TradeAmountInputField.SELL_FIAT
-        state.trade = undefined
-        state.buyAmountFiat = '0'
-        state.sellAmountFiat = '0'
-        return state
+      draft => {
+        const tradeAmountsByActionAndAmount = selectTradeAmountsByActionAndAmount(draft)
+        const sellAsset = draft.sellAsset
+        const buyAsset = draft.buyAsset
+        const {
+          sellAmountSellAssetBaseUnit,
+          buyAmountBuyAssetBaseUnit,
+          fiatSellAmount,
+          fiatBuyAmount,
+        } = tradeAmountsByActionAndAmount
+        const buyAmountCryptoPrecision =
+          buyAmountBuyAssetBaseUnit && buyAsset
+            ? fromBaseUnit(buyAmountBuyAssetBaseUnit, buyAsset.precision)
+            : '0'
+        const sellAmountCryptoPrecision =
+          sellAmountSellAssetBaseUnit && sellAsset
+            ? fromBaseUnit(sellAmountSellAssetBaseUnit, sellAsset.precision)
+            : '0'
```

Comment:
Can `bnOrZero()` + `pow pow` it and move the short circuiting inside `bnOrZero()` if you feel like it

Reply:
You no like `fromBaseUnit()`?

Reply:
Potato potato, just looked like that may oneline better 🍠